logstash-core/lib/logstash/plugins/registry.rb

```diff
@@ -0,0 +1,83 @@
+# encoding: utf-8
+require 'singleton'
+require "rubygems/package"
+
+module LogStash
+  class Registry
+
+    ##
+    # Placeholder class for registered plugins
+    ##
+    class Plugin
+      attr_reader :type, :name
+
+      def initialize(type, name)
+        @type  = type
+        @name  = name
+      end
+
+      def path
+        "logstash/#{type}s/#{name}"
+      end
+
+      def gem_name
```

Comment:
The only point this is used is to check if the plugin, coming from a non random gem name, is installed or not. 

If installed, it add the plugin to the registry himself, basically to support the old style plugins. When having more than 1 plugin in the same gem you will need somehow to register all plugins, because without any pattern there is no way to get this automatically done.

will try to simplify the logic a bit, not sure all steps are 100% necessary.